mypy/fastparse.py

```diff
@@ -190,19 +192,27 @@ def as_block(self, stmts: List[ast35.stmt], lineno: int) -> Block:
 
     def fix_function_overloads(self, stmts: List[Statement]) -> List[Statement]:
         ret = []  # type: List[Statement]
-        current_overload = []
+        current_overload = []  # type: List[Decorator]
         current_overload_name = None
         # mypy doesn't actually check that the decorator is literally @overload
         for stmt in stmts:
-            if isinstance(stmt, Decorator) and stmt.name() == current_overload_name:
+            if (isinstance(stmt, Decorator)
+                    and is_overload_part(stmt)
+                    and stmt.name() == current_overload_name):
                 current_overload.append(stmt)
+            elif (isinstance(stmt, FuncDef)
+                  and stmt.name() == current_overload_name
+                  and stmt.name() is not None):
+                ret.append(OverloadedFuncDef(current_overload, stmt))
+                current_overload = []
+                current_overload_name = None
             else:
```

Comment:
Looks like there's a missing case here: a Decorator that's not an overload but does match the current overload name. That would indicate a decorated implementation, which I think we ought to support.